### ISSUE TITLE
large numbers of playlists cause database locks

### ISSUE THREAD

**Issue by Holzhaus** [Holzhaus/mixxx-gh-issue-migration#430]
I'm having problems rescanning my library -- I get "database is locked" notifications and the scan fails.  After a couple hours of debugging I figured out how to fix it: if I delete most of my auto-generated playlists, the rescan succeeds.  I had about 500 setlog playlists in my history when I noticed this bug.

After doing a lot of testing, it seems 256  is the number at which this bug is triggered. As long as I have fewer than 256 playlists, the scan completes.  Looking more carefully, if I leave all my playlists in the database and then check out the History section of the library sidebar, I don't see all of my playlists.  I think there's a problem where we can't have more than 256 rows in the treeview?

Steps to reproduce:
1. Start up mixxx with auto rescan disabled
2. Wait until the interface has loaded
3. Click "Rescan Library"

Expected Results:
1. Successful scan

Actual Results:
1. The first DB call, m_libraryHashDao.invalidateAllDirectories();, fails with a database lock error

Workaround:
1. Using sqlite from the command line, remove a large number of playlists.  For me, I removed everything that was automatically generated:  "DELETE FROM Playlists WHERE rowid>3; DELETE FROM PlaylistTracks WHERE playlist_id>3;"

Suggested Fixes:
1. to start with, delete any playlist that has no tracks associated with it: DELETE FROM Playlists WHERE Playlists.rowid NOT IN (SELECT playlist_id FROM PlaylistTracks);
2. Figure out if this is a QT bug?

**Comment by Holzhaus** [comment]
(by rryan)
This was caused by our use of QSqlTableModel. We used to use QSTM for keeping track of tables like the main library but we ditched it because of this internal caching behavior it has. For queries with large numbers of results it keeps a lock on the database until you have read all the rows. We kept using QSqlTableModel for the Playlists and Crates table because it was worked fine but we forgot to do the usual trick of:

while (tableModel.canFetchMore()) {
  tableModel.fetchMore();
}

This tells QSTM to fully fetch the entire table instead of leaving the table locked while waiting to read the whole thing. This only appeared because the SetlogFeature creates an empty playlist every time you open Mixxx so the # of setlog playlists would slowly creep above the total # of rows that QSTM fetches at once.

Ugh.
